src/olmo_core/nn/conversion/state_mapping.py

```diff
@@ -0,0 +1,250 @@
+from dataclasses import dataclass
+from typing import Any, Dict, Optional, Set, Tuple
+
+from olmo_core.config import StrEnum
+
+
+class TemplatePlaceholder(StrEnum):
+    LAYER = "[layer]"
+    EXPERT = "[expert]"
+
+
+@dataclass
+class StateMappingTemplate:
+    """
+    The template for a mapping from state from one format to another format (e.g. OLMo Core to HF).
+    These mappings are 'templates' since they support keys and other metadata having placeholders
+    for information like the layer number or number of MoE experts. This class can be converted
+    to a `StateMapping` by providing the placeholder information information.
+
+    The most standard mapping is a one-to-one state mapping, which corresponds to a single
+    string entry for both `source_template_keys` and `dest_template_keys`. The class also supports
+    more complicated mappings, like many-to-many mappings or mappings that also require further
+    manipulations of state like permuting dimensions.
+    """
+
+    source_template_keys: str | Tuple[str, ...]
+    """
+    The key or keys of the state(s) being mapping from.
+    """
+    dest_template_keys: str | Tuple[str, ...]
+    """
+    The key or keys of the state(s) being mapping to.
+    """
+
+    source_key_per_placeholder: TemplatePlaceholder | None = None
+    """
+    A placeholder in `source_template_keys` for which this mapping should map all valid placeholder
```

Comment:
Annoyingly ReStructured Text requires double backticks for code formatting.

```suggestion
    A placeholder in ``source_template_keys`` for which this mapping should map all valid placeholder
```